src/sentry/api/endpoints/organization_member/__init__.py

```diff
@@ -1,28 +1,86 @@
 from __future__ import annotations
 
-from typing import Collection
+from typing import Collection, TypedDict
 
 from django.db import transaction
 from rest_framework.request import Request
 
 from sentry import roles
+from sentry.api.exceptions import SentryAPIException, status
 from sentry.auth.superuser import is_active_superuser
-from sentry.models import Organization, OrganizationMember, OrganizationMemberTeam
-from sentry.roles.manager import Role
+from sentry.models import Organization, OrganizationMember, OrganizationMemberTeam, Team, TeamStatus
+from sentry.roles.manager import Role, TeamRole
+
+
+class InvalidTeam(SentryAPIException):
+    status_code = status.HTTP_400_BAD_REQUEST
+    code = "invalid_team"
+    message = "The team slug does not match a team in the organization"
+
+
+class TeamRoleDict(TypedDict):
+    teamSlug: str
+    role: str
 
 
 @transaction.atomic
-def save_team_assignments(organization_member, teams):
-    # teams may be empty
+def save_team_assignments(
+    organization_member: OrganizationMember,
+    teams: list[str],
+    teams_with_roles: list[TeamRoleDict] = None,
+):
+    team_slugs = [item["teamSlug"] for item in teams_with_roles] if teams_with_roles else teams
+    target_teams = list(
+        Team.objects.filter(
+            organization=organization_member.organization,
+            status=TeamStatus.VISIBLE,
+            slug__in=team_slugs,
+        )
+    )
+    if len(target_teams) != len(team_slugs):
+        raise InvalidTeam
```

Comment:
This could lead to an awkward failure mode if there is a duplicate slug, which I think it possible if the original REST request lists one twice. Maybe put `team_slugs` into a `set` to fix that.

It also would be pretty easy to see which team slug was unmatched (`set(team_slugs).difference(set(team.slug for team in target_teams))`) if we want to include that in the error message, but maybe that level of detail isn't typical for our API responses.